### PR TITLE
Fix collision on `--closure 1 -s MODULARIZE=1 -s EXPORT_NAME="HelloWorld"`

### LINKED ISSUE #4538: The result of -O2 --closure 1 -s MODULARIZE=1 is broken
Because `MODULARIZE` assumes that the following code uses name `Module`, when `closure` and `O2` are specified, it is no the case.

```
emcc -O2 --closure 1 -s MODULARIZE=1 main.c
```

main.c

```
#include <emscripten.h>

int main(int argc, char *argv[]) {
  printf("hello\n");
  return 0;
}
```

output js:

```
var Module = function(Module) {
  Module = Module || {};

var b;b||(b=eval("(function() { try { return Module || {} } catch(e) { return {} } })()"));var g={},l;for(l in b)b.hasOwnProperty(l)&&(g[l]=b[l]);var n=!1,p=!1,r=!1,t=!1;
...
...

  return Module;
};
```

Because `Module` and `b` differ, the function returns empty object. This behavior is inconsistent the output without `-O2` or `-closure 1`. The problem may be more clear with `-s EXPORT_NAME'"exportName"'`.

### PR DESCRIPTION
**Pull request by SaschaNaz**
Fixes #4538.

Current:

```js
var test = function(Module) {
  Module = Module || {};

var d;d||(d=eval("(function() { try { return test || {} } catch(e) { return {} } })()"));
/* ... */

  return Module;
};
```

Fix:

```js
var test = function(Module) {
  Module = Module || {};

var d;d||(d=eval("(function() { try { return Module || {} } catch(e) { return {} } })()"));
/* ... */

  return Module;
};
```

### CODE REVIEW

**Comment by kripken** [comment]
Not sure I follow what this fixes - the issue it came from had some examples, but some used embind, so I'm not sure there's a three-way collision here?

Please add a simple testcase to this pull, which we need anyhow for the test suite, and can also help explain the problem.

**Comment by SaschaNaz** [comment]
Tried adding a test but couldn't really run it because of this issue:

```
C:/Program Files/Emscripten/clang/fastcomp/build_incoming_vs2013_64/RelWithDebInfo/bin\llvm-link: c:/users/sasch/appdata/local/temptmpe74iohstrlwr.c.o: error: Could not open input file: no such file or directory
C:/Program Files/Emscripten/clang/fastcomp/build_incoming_vs2013_64/RelWithDebInfo/bin\llvm-link: error loading file 'c:/users/sasch/appdata/local/temptmpe74iohstrlwr.c.o'
Traceback (most recent call last):
  File "C:\Users\sasch\Documents\GitHub\emscripten\emcc", line 13, in <module>
    emcc.run()
  File "C:\Users\sasch\Documents\GitHub\emscripten\emcc.py", line 1280, in run
    extra_files_to_link += system_libs.calculate([f for _, f in sorted(temp_files)] + extra_files_to_link, in_temp, stdout, stderr, forced=forced_stdlibs)
  File "C:\Users\sasch\Documents\GitHub\emscripten\tools\system_libs.py", line 356, in calculate
    libfile = shared.Cache.get(name, do_create, extension=suffix)
  File "C:\Users\sasch\Documents\GitHub\emscripten\tools\cache.py", line 41, in get
    temp = creator()
  File "C:\Users\sasch\Documents\GitHub\emscripten\tools\system_libs.py", line 354, in do_create
    ret = create(name)
  File "C:\Users\sasch\Documents\GitHub\emscripten\tools\system_libs.py", line 120, in create_libc
    return build_libc(libname, libc_files, args)
  File "C:\Users\sasch\Documents\GitHub\emscripten\tools\system_libs.py", line 67, in build_libc
    shared.Building.link(o_s, in_temp(lib_filename))
  File "C:\Users\sasch\Documents\GitHub\emscripten\tools\shared.py", line 1449, in link
    assert os.path.exists(target) and (output is None or 'Could not open input file' not in output), 'Linking error: ' + output
AssertionError: Linking error:
[browser launch: a.html ]
ERROR
[Browser harness server terminated]
```

I can't understand why the test tries to find a file on my `appdata/local` directory.

**Comment by SaschaNaz** [comment]
BTW, It seems`-s EXPORT_NAME` does no help without `-s MODULARIZE=1`, is this the intended behavior? If so, I think the line without closure should be also changed to `Module` instead of `{{{ EXPORT_NAME }}}`.

**Comment by kripken** [comment]
From the path I think it's trying to create a temp file. I'm not sure what `appdata/local` is, perhaps that is what python reports as the temp dir on your system? You can override the temp dir by setting the value of `TEMP_DIR` in `~/.emscripten`.

Is this problem only when using closure? Does other stuff work?

`EXPORT_NAME` is mostly used with `MODULARIZE`, yes. But also when we export the module in node, and it is the name we look for in the global scope to write to (which is why we use it in the non-closure version of that line).

**Comment by SaschaNaz** [comment]
Not sure what 'this problem' means here. If you mean the collision, yes it works without closure option. If you mean the test, no it fails with every test on `python tests/runner.py browser`.

**Comment by kripken** [comment]
I looked into this locally to try to get a better idea of what's going on. On the incoming branch, the test passes with your modifications - so if it doesn't pass for you, perhaps you were testing on master, and this is something that was fixed since then? Or perhaps the general problem you have with tests is interfering? (probably worth figuring that out - maybe pick one of the failing test, and run it with `EMCC_DEBUG=1` in the env, that might show something helpful)

In other words, we can merge your `tests/` change, without the `src/` change. Which makes sense to do, we should have test coverage for this.

Or am I still missing something here?

**Comment by SaschaNaz** [comment]
Hi, sorry to respond late.

I cannot understand how the test passed, as it should always fail by my mistake. The `Module` identifier should be always `undefined` on the new test line because of earlier `if (typeof Module !== "undefined")` check...

**Comment by juj** [comment]
In the above paste snippet, the line

    C:/Program Files/Emscripten/clang/fastcomp/build_incoming_vs2013_64/RelWithDebInfo/bin\llvm-link: error loading file 'c:/users/sasch/appdata/local/temptmpe74iohstrlwr.c.o'

suggests that something is going wrong with escaping backslashes. The expected filename for this is `c:/users/sasch/appdata/local/temp/tmpe74iohstrlwr.c.o`. I vaguely remember having fixed something like this somewhat recently in the incoming branch.

Another thing that looks odd is that the current incoming branch should require Visual Studio 2015 Update 3 and it should not build anymore with VS2013, because upstream LLVM started requiring VS2015 to build. The directory name 'vs2013' here looks suspect, it should be 'vs2015' now. Can you try explicitly choosing vs2015 to build and activate emsdk with, i.e.

    emsdk install --vs2015 sdk-incoming-64bit
    emsdk activate --vs2015 sdk-incoming-64bit

That should generate fastcomp in a vs2015 build directory. I don't know if this affects the root issue with the tests here, but there is a chance that the content in `build_incoming_vs2013_64` is not at all up to date.

**Comment by SaschaNaz** [comment]
@juj Thanks for the tip! Unfortunately I'm still getting the same error with `--vs2015`. Not sure why this only happens when running test.

```
C:/Program Files/Emscripten/clang/fastcomp/build_incoming_vs2015_64/RelWithDebInfo/bin\llvm-link: c:/users/sasch/appdata/local/temptmp
yx3d2jstrlwr.c.o: error: Could not open input file: no such file or directory
C:/Program Files/Emscripten/clang/fastcomp/build_incoming_vs2015_64/RelWithDebInfo/bin\llvm-link: error loading file 'c:/users/sasch/a
ppdata/local/temptmpyx3d2jstrlwr.c.o'
Traceback (most recent call last):
  File "C:\users\sasch\documents\github\emscripten\emcc", line 13, in <module>
    emcc.run()
  File "C:\users\sasch\documents\github\emscripten\emcc.py", line 1280, in run
    extra_files_to_link += system_libs.calculate([f for _, f in sorted(temp_files)] + extra_files_to_link, in_temp, stdout, stderr, fo
rced=forced_stdlibs)
  File "C:\users\sasch\documents\github\emscripten\tools\system_libs.py", line 356, in calculate
    libfile = shared.Cache.get(name, do_create, extension=suffix)
  File "C:\users\sasch\documents\github\emscripten\tools\cache.py", line 41, in get
    temp = creator()
  File "C:\users\sasch\documents\github\emscripten\tools\system_libs.py", line 354, in do_create
    ret = create(name)
  File "C:\users\sasch\documents\github\emscripten\tools\system_libs.py", line 120, in create_libc
    return build_libc(libname, libc_files, args)
  File "C:\users\sasch\documents\github\emscripten\tools\system_libs.py", line 67, in build_libc
    shared.Building.link(o_s, in_temp(lib_filename))
  File "C:\users\sasch\documents\github\emscripten\tools\shared.py", line 1449, in link
    assert os.path.exists(target) and (output is None or 'Could not open input file' not in output), 'Linking error: ' + output
AssertionError: Linking error:
[browser launch: page.html ]
ERROR
```

**Comment by juj** [comment]
Hmm, can you paste the contents of `C:\users\sasch\.emscripten`? Perhaps that would give a clue.

**Comment by juj** [comment]
Oh, also, pull requests should go to `incoming` branch, but this is against `master`.

**Comment by SaschaNaz** [comment]
Aw, this PR should be closed then. I opened a new issue #4570 for this.
